enumerable.rb

```diff
@@ -0,0 +1,143 @@
+require_relative 'helper_methods'
+
+module Enumerable
+  def my_each()
+    return to_enum(:my_each) unless block_given?
+
+    arr = *self # if a range is given, it splats it into an array
+    size.times do |indx|
+      yield arr[indx]
+    end
+    self
+  end
+
+  def my_each_with_index
+    return to_enum(:my_each) unless block_given?
+
+    arr = *self # if a range is given, it splats it into an array
+    size.times do |indx|
+      yield arr[indx], indx
+    end
+    self
+  end
+
+  def my_select
+    return to_enum(:my_each) unless block_given?
+
+    selected_elements = []
+    my_each do |k|
+      selected_elements.push(k) if yield k
+    end
+    selected_elements
+  end
+
+  def my_all?(*param)
+    return Helper.match_all(self, param[0]) if param.length.positive?
+
+    # returns false if a block is not given and falsy element is found
+    my_each { |ele| return false unless ele || block_given? }
+    # returns true if a block is not given and falsy element is not found
+    return true unless block_given?
+
+    my_each do |ele|
+      # returns false if an element that doesn't satsify the condition is found
+      return false unless yield ele
+    end
+    true
+  end
+
+  def my_any?(*param)
+    return Helper.match_any(self, param[0]) if param.length.positive?
+
+    # returns true if a block is not given and truthy element is found
+    my_each { |ele| return true if ele && !block_given? }
+    # returns false if a block is not given and truthy element is not found
+    return false unless block_given?
+
+    my_each do |ele|
+      # returns true if an element that satsifys the condition is found
+      return true if yield ele
+    end
+    false
+  end
+
+  def my_none?(*param)
+    return Helper.match_none(self, param[0]) if param.length.positive?
+
+    # returns true if a block is not given and truthy element is found
+    my_each { |ele| return false if ele && !block_given? }
+    # returns false if a block is not given and truthy element is not found
+    return true unless block_given?
+
+    my_each do |ele|
+      # returns false if an element that satsifys the condition is found
+      return false if yield ele
+    end
+    true
+  end
```

Comment:
[OPTIONAL]

I think we could get the block of my_none? from my_any? we should only do something like !my_any? for my_none? to work. For example:

```ruby
def my_none?(arg = nil, &block)
  !my_any?(arg, &block)
end
```

To make this work we have to use block.call(element) instead of yield(element) in my_any?